### DISCUSSION TITLE
Help with periodic boundary conditions

### DISCUSSION THREAD

**Discussion by 10UAikawa** [NanoComp/meep#2681] [Q&A] [answered]
So, I noticed in my simulation that when I have my Gaussian source emit, that it will transmit through all the way to the other side and it looks like this is due to my periodic boundary conditions, I also noticed that I could move my source to a position outside of the unit cell scope for example set it to "center=mp.Vector3(999999,0, ((Z_size/2)-100) )" and it would move my source by some position. And this is because it is periodic. I think the same thing is happening with the radiation because it is moving to the other side. And I dont what my system to be periodic in the z-direction. 

Here is an image of my problem:
![image](https://github.com/NanoComp/meep/assets/71509442/1e6791f7-c034-48b8-93ec-379927b77fd7)
Notice how it is emitting on the other side. How can I stop this?

Here is my code for reference:

import meep as mp
import matplotlib.pyplot as plt
import numpy as np

'''-----units are in microns (1 MEEP unit = 1 um)-----'''
'''-----with time scale as t=3.3e-15s since a/c gives time scale-----'''

#pixels per mircon (pixel/um)
resolution = 0.1

#Padding
dpml = 1

#size of the cell
X_size = 420
Y_size = 420
Z_size = 800

#dimensions of the quartz block
x_block = 400
y_block = 400
z_block = 100

#dimensions of the ribbon 
ribbon_width = 230
rib_thickness = 0.02

#dimensions of the aluminum
alum_slit_spacing = 344
alum_slit_width = 5   
alum_slit_height = 220
alum_thickness = 0.02

"""Need to figure out how to correctly convert units and permitvity/conductivty"""
#Materials
Cd3As2_conductivity = 1e5  # S/m
Cd3As2_epsilon = 1  # relative permitvity is already dimensionless (1 for metals/semimetals)
Al_conductivity = 38000000 # S/m
Al_epsilon = 1.000021 # relative permitvity is already dimensionless (1 for metals/semimetals)
Quartz_epsilon = 4.41  # relative permitvity is already dimensionless 
freq_intrest = 0.4 #freq at 0.4THz it is consistent with MEEP units (since 1=1um)
wavelength = 749.481 # wavelength of frequency at 0.4THz

# conversion = σD=(a/c)σ/(εrε0)
eo=8.8541878128E-12  #in F/m
a=1e-6  #units in meters from micron to meters (since we use micron units as 1)
c=3e8  #in SI units (m/s)
er=1

#Using a constant conductivty because e are examining this a narrow frequency band so 
#we can use this as an approximtion (Also this same method as HFSS to compare)
Cd3As2_conductivity_MEEP_units = (a/c)*Cd3As2_conductivity/(er*eo)
Al_conductivity_MEEP_units = (a/c)*Al_conductivity/(Al_epsilon*eo)
print("MEEP units: ",Al_conductivity_MEEP_units)
 
# Create a material with the specified conductivity/permitivty
Cd3As2_material = mp.Medium(epsilon = Cd3As2_epsilon, D_conductivity = (2*np.pi)*freq_intrest*Cd3As2_conductivity_MEEP_units/Cd3As2_epsilon )
Al_material = mp.Medium(epsilon = Al_epsilon, D_conductivity = (2*np.pi)*freq_intrest*Al_conductivity_MEEP_units/Al_epsilon)
Quartz_material = mp.Medium(epsilon = Quartz_epsilon)

#defining cell size
cell = mp.Vector3(X_size, Y_size, Z_size)

#Bloch-periodic boundary conditions
k_point = mp.Vector3(x_block/2, y_block/2, 0) #because periodic in the x and y direction 
# k_point = mp.Vector3()

#frequnecy of intrest and range
fcen = 0.4  # pulse center frequency 0.4 THz
df = 0.3  # pulse width (in frequency)

#boundary conditions
pml_layers = [mp.PML(dpml)]

geometry_lattice = mp.Lattice(size=mp.Vector3(X_size,Y_size,Z_size))

'''To create geometry need: 
-7 aluminum blocks
-4 quartz blocks
-1 Cd3As2 block'''

""" NOTICE: only beginning and end should have z_length/2 """

'''EXAMPLE: Cd3As2 Ribbon (move up z-dir by (z_block/2) because 
that is the center of the first quartz block because 100/2=50 
and that is the center part of the block)'''

geometry = [
    #Quartz (bottom)
    mp.Block(
        mp.Vector3(x_block, y_block, z_block),
        center=mp.Vector3(0,0,0),
        material=Quartz_material,
    ),

    #Quartz (middle quartz 1)
    mp.Block(
        mp.Vector3((x_block-ribbon_width)/2, y_block, rib_thickness),
        center=mp.Vector3((ribbon_width/2)+(y_block-ribbon_width)/4  , 0 ,  (z_block/2)+(rib_thickness/2)),
        material=Quartz_material,
    ),

    #Quartz (middle quartz 2)
    mp.Block(
        mp.Vector3((x_block-ribbon_width)/2, y_block, rib_thickness),
        center=mp.Vector3( -((ribbon_width/2)+((y_block-ribbon_width)/4))  , 0 ,  (z_block/2)+(rib_thickness/2)),
        material=Quartz_material,
    ),

    #Cd3As2 Ribbon (middle)
    mp.Block(
        mp.Vector3(ribbon_width, y_block, rib_thickness),
        center=mp.Vector3(0,0,(z_block/2)+(rib_thickness/2)),
        material=Cd3As2_material,
    ),

    #Quartz (top)
    mp.Block(
        mp.Vector3(x_block, y_block, z_block),
        center=mp.Vector3(0,0,(z_block/2)+(rib_thickness)+(z_block/2)),
        material=Quartz_material,
    ),

    #Alumninum (Block 1)
    mp.Block(
        mp.Vector3( (x_block-(alum_slit_width*2 + alum_slit_spacing))/2 , y_block, alum_thickness),
        center=mp.Vector3((-x_block/2) + ((x_block-(alum_slit_width*2 + alum_slit_spacing))/2)/2 ,0,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    ),

    #Alumninum (Block 7)
    mp.Block(
        mp.Vector3( (x_block-(alum_slit_width*2 + alum_slit_spacing))/2 , y_block, alum_thickness),
        center=mp.Vector3((x_block/2) - ((x_block-(alum_slit_width*2 + alum_slit_spacing))/2)/2 ,0,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    ),

    #Alumninum (Block 2)
    mp.Block(
        mp.Vector3(alum_slit_width, (y_block - alum_slit_height)/2 , alum_thickness),
        center=mp.Vector3( -((alum_slit_spacing/2) + alum_slit_width/2) , -(y_block/2 - ((y_block - alum_slit_height)/4)) ,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    ),

    #Alumninum (Block 3)
    mp.Block(
        mp.Vector3(alum_slit_width, (y_block - alum_slit_height)/2 , alum_thickness),
        center=mp.Vector3( -((alum_slit_spacing/2) + alum_slit_width/2) , y_block/2 - ((y_block - alum_slit_height)/4) ,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    ),

    #Alumninum (Block 5)
    mp.Block(
        mp.Vector3(alum_slit_width, (y_block - alum_slit_height)/2 , alum_thickness),
        center=mp.Vector3( ((alum_slit_spacing/2) + alum_slit_width/2) , -(y_block/2 - ((y_block - alum_slit_height)/4)) ,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    ),

    #Alumninum (Block 6)
    mp.Block(
        mp.Vector3(alum_slit_width, (y_block - alum_slit_height)/2 , alum_thickness),
        center=mp.Vector3( ((alum_slit_spacing/2) + alum_slit_width/2) , (y_block/2 - ((y_block - alum_slit_height)/4)) ,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    ),

    #Alumninum (Block 4)
    mp.Block(
        mp.Vector3(alum_slit_spacing, y_block, alum_thickness),
        center=mp.Vector3(0,0,(z_block/2)+(rib_thickness)+(z_block)+(alum_thickness/2)),
        material=Al_material,
    )
]

'''its a good idea to give enought space between PML and source so you dont get reflections'''
#setting up the source 
sources = [
    mp.Source(
        mp.GaussianSource(fcen, fwidth=df),
        component=mp.Ez,
        center=mp.Vector3(0,0, ((Z_size/2)-100) ),
        size=mp.Vector3(x_block, y_block, 0),
    )
]

#setting up the simulation
sim = mp.Simulation(
    cell_size=cell,
    boundary_layers=pml_layers,
    geometry=geometry,
    k_point=k_point,
    sources=sources,
    resolution=resolution,
)

sim.run(until=200)
#getting data from the array
ez_data = sim.get_array(center=mp.Vector3(), size=cell, component=mp.Ey)
# Choose the z-slice (e.g., z_index) you want to visualize
index_slice = 0
# Extract the 2D slice in the x-y plane at the chosen z-slice
ez_slice = np.real(ez_data[:, index_slice,:])
print("min: ",np.min(ez_slice),"    ","min: ",np.max(ez_slice) )

'''-------------This is how I can view a slice of a 3d object in a 2d plane------------------'''
# Define the slice plane
slice_pos = (z_block/2)+(rib_thickness)+(z_block)+(alum_thickness+14)  # Position of the slice along the z-axis
# slice_dim = mp.Vector3(0, 420, 800)  # Dimensions of the slice (change (x,y,z) x=0 to see in y,z plane)

slice_dim = mp.Vector3(0, 420, 800) # y-z plane
# slice_dim = mp.Vector3(420, 0, 800) # x-z plane
# slice_dim = mp.Vector3(420, 420, 0) # x-y plane

# Get the Dielectric data (for example, the Dielectric component to see the geometry)
Die_data = sim.get_array(center=mp.Vector3(0, 0, slice_pos), size=slice_dim, component=mp.Dielectric)

# Plot the 2D slice using Matplotlib
plt.figure()
plt.imshow(Die_data.transpose(), interpolation='spline36', cmap='binary')
plt.imshow(ez_slice.transpose(), interpolation='spline36', cmap='RdBu', alpha=0.7)
plt.axis('on')
plt.show()
'''------------------------------------------------------------------------------------------'''

**Comment by smartalecH** [comment, marked as answer]
> And I dont what my system to be periodic in the z-direction.

Just add PML or an adiabatic absorber in the z direction.
